src/storage/store/var_list_column_chunk.cpp

```diff
@@ -109,83 +139,113 @@ void VarListColumnChunk::lookup(
+    KU_ASSERT(sizeColumnChunk->sanityCheck());
 }
 
 void VarListColumnChunk::write(
     ColumnChunk* chunk, ColumnChunk* dstOffsets, RelMultiplicity /*multiplicity*/) {
-    needFinalize = true;
-    if (!indicesColumnChunk) {
-        initializeIndices();
-    }
     KU_ASSERT(chunk->getDataType().getPhysicalType() == dataType.getPhysicalType() &&
               dstOffsets->getDataType().getPhysicalType() == PhysicalTypeID::INT64 &&
               chunk->getNumValues() == dstOffsets->getNumValues());
-    auto currentIndex = numValues;
-    append(chunk, 0, chunk->getNumValues());
+    checkOrder = false;
+    offset_t currentIndex = varListDataColumnChunk->getNumValues();
+    auto otherListChunk = ku_dynamic_cast<ColumnChunk*, VarListColumnChunk*>(chunk);
+    varListDataColumnChunk->resizeBuffer(varListDataColumnChunk->getNumValues() +
+                                         otherListChunk->varListDataColumnChunk->getNumValues());
+    varListDataColumnChunk->dataColumnChunk->append(
+        otherListChunk->varListDataColumnChunk->dataColumnChunk.get(), 0,
+        otherListChunk->varListDataColumnChunk->getNumValues());
+    offset_t maxDstOffset = 0;
     for (auto i = 0u; i < dstOffsets->getNumValues(); i++) {
         auto posInChunk = dstOffsets->getValue<offset_t>(i);
-        KU_ASSERT(posInChunk < capacity);
-        indicesColumnChunk->setValue<int64_t>(currentIndex++, posInChunk);
-        indicesColumnChunk->getNullChunk()->setNull(posInChunk, false);
-        if (indicesColumnChunk->getNumValues() <= posInChunk) {
-            indicesColumnChunk->setNumValues(posInChunk + 1);
+        if (posInChunk > maxDstOffset) {
+            maxDstOffset = posInChunk;
         }
     }
-    KU_ASSERT(currentIndex == numValues &&
-              indicesColumnChunk->getNumValues() <= indicesColumnChunk->getCapacity());
+    while (maxDstOffset >= numValues) {
+        appendNullList();
+    }
+    for (auto i = 0u; i < dstOffsets->getNumValues(); i++) {
+        auto posInChunk = dstOffsets->getValue<offset_t>(i);
+        auto appendLen = otherListChunk->getListLen(i);
+        currentIndex += appendLen;
+        setValue<offset_t>(currentIndex, posInChunk);
+        nullChunk->setNull(posInChunk, otherListChunk->nullChunk->isNull(i));
+        sizeColumnChunk->setValue<uint32_t>(appendLen, posInChunk);
+        sizeColumnChunk->getNullChunk()->setNull(posInChunk, otherListChunk->nullChunk->isNull(i));
+    }
+    KU_ASSERT(sizeColumnChunk->getNumValues() == numValues);
+    KU_ASSERT(nullChunk->getNumValues() == numValues);
+    KU_ASSERT(sizeColumnChunk->getNullChunk()->getNumValues() == numValues);
+    KU_ASSERT(sizeColumnChunk->sanityCheck());
 }
 
 void VarListColumnChunk::write(
     ValueVector* vector, offset_t offsetInVector, offset_t offsetInChunk) {
-    needFinalize = true;
-    if (!indicesColumnChunk) {
-        initializeIndices();
+    checkOrder = false;
```

Comment:
Why not set `checkOrder` to true when we have writes? so we need to "check order" when it is set to true.